Add Test Client and Framework for e2e tests

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
This PR add a `TestClient` and Framework. The `TestClient` is a wrapped kubevirt client that exposes all client functionality and can be used as a normal client. In addition it provides a mechanism that track the non-namespaces resources created or updated during e2e tests, and functions to rollback (cleanup) them at the end of the execution.
This` Framework` supports common operations used by e2e tests; it will keep a client & clean the cluster for you. It is not necessary anymore to clean the cluster on `BeforeEach` of all tests if you use the framework because it cleans it for you. It contains a call to `tests.BeforeTestCleanup()` on its `BeforeEach()`. 
It also initializes the `TestClient` that can be used everywhere: in this way, writing tests, you have not to retrieve the `TestClient` because it creates one for you, you can simply use it. 
Looking at the tests files there where different cases where`tests.BeforeTestCleanup()` was defined inside `BeforeEach()` of the outer `Describe` block and also in the inner ones(`Context, It, Describe`..). Since `BeforeEach()` is executed for each specs it causes a multiple calls to `tests.BeforeTestCleanup()` function.
In other cases, it was not defined in any of the `BeforeEach()` causing a possible dependency between tests.
The union of these two gives the following advantages:
1. Make sure that each test will run in a cleaned cluster
2. Provides to you a `TestClient` that can be used without the necessity of retrieving ones
3. Gives the possibility to write a `BeforeEach()` that is specific for that set of tests

Since it cleans the cluster before each test this ensures that every test runs in the same cluster conditions.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Only a subset of the tests is affected by changes in order to monitoring issues.
The tests were chosen on the basis of whether or not they contained changes to non-namespaces resources tracked by the test client:

- perstistenVolume CREATE
- node CREATE, UPDATE, PATCH
- namespace CREATE
- migrationPolicy CREATE

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add TestClient to easily rollback non-namespaced resources created or updated during e2e tests
Add Framework to support common operations used by functional tests
```

fossedihelm commented:
/cc @rmohr

fossedihelm commented:
/test all

fossedihelm commented:
/test pull-kubevirt-e2e-k8s-1.23-sig-network

fossedihelm commented:
/test pull-kubevirt-unit-test

fossedihelm commented:
/retest

fossedihelm commented:
/test all

fossedihelm commented:
/cc @xpivarc

fossedihelm commented:
/test all

fossedihelm commented:
/ok-to-test

fossedihelm commented:
/retest

fossedihelm commented:
> I just had quick look and I think it would help me if we split the changes to tests from the first commit. It also is logical imho. WDYT?

@xpivarc Thanks, you are definitely right. Will do

fossedihelm commented on tests/framework/client/persistentvolume.go:
```diff
@@ -0,0 +1,93 @@
+	"k8s.io/client-go/rest"
+)
+
+var createdPVs = make([]pvInfo, 0)
```
The idea was to keep each "tracked" information in its own context, but if you think that keep it all in the client will increase the readability, or for some other reasons, I have no problem to change it :)

xpivarc replied:
I think this is fine as far we don't use client concurently, right?

fossedihelm replied:
The idea is not to use another `client` concurrently because otherwise the `testClient` is bypassed. This has to be forethought of the developers and the reviewers. On the other hand, also if you use the `kubectl` or the `RestClient` you can bypass it.

fossedihelm commented:
@xpivarc I arrange the commits in the following order:
- Framework creation
- TestClient creation
- Changes to e2e tests

fossedihelm commented:
/test pull-kubevirt-e2e-k8s-1.23-sig-storage-cgroupsv2

fossedihelm commented:
/test pull-kubevirt-e2e-k8s-1.23-sig-storage-cgroupsv2

fossedihelm commented:
/retest

fossedihelm commented:
/retest

fossedihelm commented:
/retest

xpivarc commented on tests/framework/client/node.go:
```diff
@@ -0,0 +1,134 @@
+			panic(err)
+		}
+	}
+	for name, node := range nodeSpecMap {
```
We could also check if we need to update the node.

fossedihelm replied:
I add a check on clean as you suggested, thanks

xpivarc commented on tests/framework/client/node.go:
```diff
@@ -0,0 +1,134 @@
+
+func (c *nodes) saveNodeSpecs(ctx context.Context, name string) {
+	//check that changes are not referred to a previous created node
+	if _, exist := createdNodes[name]; !exist {
```
I would probably negate this. If we are changing created node for a test we don't need to keep the previous state.

fossedihelm replied:
Probably the comment is misleading. Here we check that the node is gonna to be modified is not a previous created one:
if we want to patch a node `x` and `x` was manually created during the test I'm not going to save the state of `x` because I know that `x` will be deleted at the end of the test.

xpivarc replied:
Exactly I was just suggesting doing `if _, exist := createdNodes[name]; exist { //node will be deleted return}`. Just suggestion.

fossedihelm replied:
Oh I got it. Sorry, my bad. Agree

fossedihelm replied:
Done

xpivarc commented on tests/framework/client/node.go:
```diff
@@ -0,0 +1,134 @@
+			util.PanicOnError(err)
+			nodeSpecMap[currentNode.Name] = nodeSpec{
+				labels: currentNode.Labels,
+				spec:   currentNode.Spec,
```
Do we modify the spec of nodes currently?

fossedihelm replied:
In the `Spec` we have `Unschedulable` and `Taints` fields, and in some tests we modify them. Instead of store the specific fields I opted to store all the `Spec`

fossedihelm replied:
I modified the `nodeSpec` struct, using `Unschedulable` and `Taints` directly, avoiding storing other stuff we are not interested in. Thanks

fossedihelm commented on tests/framework/framework/framework.go:
```diff
@@ -0,0 +1,111 @@
+type Framework struct {
+	BaseName string
+
+	UniqueName string
```
No, you didn't miss anything. Currently it is never used. I think that it could be useful for future improvements (i.g. create test specific labels). I can delete it if you think that is better to "add it when we need it"

fossedihelm commented on tests/framework/framework/framework.go:
```diff
@@ -0,0 +1,111 @@
+		f.clientConfig = config
+	}
+	f.UniqueName = fmt.Sprintf("%s-%08x", f.BaseName, rand.Int31())
+	tests.BeforeTestCleanup()
```
Sorry I don't understand. This function call makes the cluster clean before all tests, instead of declare it in each single test file. Why do we have to remove it?

brybacki replied:
Is there a way to design it so failed test reporting (with log dump) runs after tests, and cleanup is done later, after that? I suppose this should work, because we use `JustAfterEach` for reporting and we can use `AfterEach` for cleanup. 

Is there a problem that forces us to clean before test? Is it better to clean before?

brybacki replied:
I have started reading a discussion on kubevirt-dev. I can see it mentioned as "an intermediate solution". So maybe I need to finish reading the discussion before reviewing this PR.

fossedihelm replied:
@brybacki As you said we collect reporting data in the `JustAfterEach` so yes, basically we can cleanup in the `AfterEach`.
Right, the `Framework` was intended, at least by me, as an intermediate solution that should be extended step by step introducing functionalities. But it's ok for me to drop it from this PR and discuss which functionalities we want in the proposal design.

fossedihelm commented on tests/framework/client/client.go:
```diff
@@ -0,0 +1,70 @@
+
+var (
+	coreOnce                  sync.Once
+	testCoreV1Client          *TestCoreV1Client
```
In the client we initialize the `coreV1Client` during [https://github.com/kubevirt/kubevirt/blob/a6a84d0388b12d8c812e19cb7cc88f5193e42bdf/staging/src/kubevirt.io/client-go/kubecli/kubecli.go#L146](https://github.com/kubevirt/kubevirt/blob/a6a84d0388b12d8c812e19cb7cc88f5193e42bdf/staging/src/kubevirt.io/client-go/kubecli/kubecli.go#L146) and it is done only once (because `CoreV1()` is managed by k8s itself)
I don't think it is necessary do it in the client itself, because it is related to the fact that I have to override the `CoreV1()` function. I don't know if this reply to your question.

fossedihelm commented:
@xpivarc Thanks for your review. I reply to your comments. Thanks

fossedihelm commented:
@xpivarc Pushed requested changes and edited some `tests/utils.go` functions: passing `virtClient` as parameter we ensure to not bypass non-namespaces resources creation. Thanks

fossedihelm commented:
I edited last commit in order to start with a subset of the tests to iterate on possible issues.
The tests were chosen on the basis of whether or not they contained changes to non-namespaces resources tracked by the test client:

- perstistenVolume CREATE
- node CREATE, UPDATE, PATCH
- namespace CREATE
- migrationPolicy CREATE

fossedihelm commented:
/retest

fossedihelm commented:
/retest

iholder-redhat commented:
/cc @iholder-redhat

iholder-redhat commented on tests/framework/client/client.go:
```diff
@@ -0,0 +1,70 @@
+	"kubevirt.io/client-go/kubecli"
+)
+
+// TestClient
+//This client is a wrapped kubevirt client that exposes all client functionality
+//and can be used as a normal client. It provides a mechanism that track
+//non-namespaced resources created or updated during e2e tests and functions to rollback them at the end of the execution.
+
+type TestClient struct {
```
nit: would change to:
```go
// TestClient is a wrapped kubevirt client that exposes all client functionality
// and can be used as a normal client. It provides a mechanism that track
// non-namespaced resources created or updated during e2e tests and functions to rollback them at the end of the execution.
type TestClient struct {
```

that is:
- "TestClient" is now part of the explanation
- added space between slashes and text
- removed empty line between doc and struct declarazation

fossedihelm replied:
Great, thanks :+1:

iholder-redhat commented on tests/framework/client/client.go:
```diff
@@ -0,0 +1,70 @@
+	return testCoreV1Client
+}
+
+//
```
nit: leftover

iholder-redhat commented on tests/framework/client/migrationpolicy.go:
```diff
@@ -0,0 +1,85 @@
+// migrationPolicies implements MigrationPolicyInterface
+type migrationPolicies struct {
+	client          rest.Interface
+	migrationPolicy migrationsv1.MigrationPolicyInterface
```
To avoid having to re-implement all of the interface's, you better [embed](https://travix.io/type-embedding-in-go-ba40dd4264df) the migration policy client.

In other words, if you'd define the struct that way:
```go
type migrationPolicies struct {
	migrationsv1.MigrationPolicyInterface
	client          rest.Interface
}

// newMigrationPolicies returns a MigrationPolicies
func newMigrationPolicies(c *TestMigrationPolicyClient) *migrationPolicies {
	return &migrationPolicies{
		MigrationPolicyInterface: c.client.MigrationPolicies(),
		client:                   c.client.RESTClient(),
	}
}
```

Then you would be able to simply:
```go
var c *migrationPolicies = ...
c.Get()
```

And you wouldn't need to recreate all Get, List, Create, etc methods.

This is also easier to implement but would make sure we're up to date with client's methods :)

iholder-redhat replied:
p.s. you can still override whatever methods you need, like Create()
and obviously the above comment applies to every client your created, not just in this file

fossedihelm replied:
You're totally right. I had some troubles trying to do that, but now it works. Thank you so much

iholder-redhat commented on tests/framework/client/migrationpolicy.go:
```diff
@@ -0,0 +1,85 @@
+	return c.migrationPolicy.UpdateStatus(ctx, migrationPolicy, opts)
+}
+
+func (c *migrationPolicies) Delete(ctx context.Context, name string, opts v1.DeleteOptions) error {
```
Maybe in delete we would like to take the policy out of `createdMigrationPolicies`?
But on the other hand this is just a pointer and shouldn't really effect test performance / usa a lot of memory

fossedihelm replied:
We could, but since  `Clean()` ignore notFound errors, I think we can leave it as is. At most we try to remove it twice

iholder-redhat replied:
Right, but it still has two disadvantages:
* Memory leak (although very small since we're only dealing with pointers)
* Overhead (since you make an API call for nothing)

Of course these two are far from being significant. Your call :)

fossedihelm replied:
Agreed, In this case I think I will move `createdMigrationPolicies` from slice to map in order to not iterate over it for every `Delete`. WDYT?

iholder-redhat replied:
Sounds good to me :+1: 
I would also remove it from the map if I know a policy is already deleted

iholder-redhat commented on tests/framework/client/client.go:
```diff
@@ -0,0 +1,70 @@
+	return testMigrationPolicyClient.MigrationPolicies()
+}
+
+func (c *TestClient) Clean() {
```
What I'd suggest here is to have a `[]CleanableResource` slice defined with the vars above.
Then, when each client is created (e.g. when calling `MigrationPolicy()`) you can use the `sync.One` to append the client to this slice.

Finally when you get here, you can simply iterate the slice and call each element's "Clean()" method.

fossedihelm replied:
Thanks, I like it

iholder-redhat commented on tests/framework/framework/framework.go:
```diff
@@ -0,0 +1,114 @@
+
+	// run all aftereach functions in random order to ensure no dependencies grow
+	for _, afterEachFn := range f.afterEaches {
+		afterEachFn(f, CurrentGinkgoTestDescription().Failed)
```
My IDE says:
```
Deprecated: CurrentGinkgoTestDescription has been replaced with CurrentSpecReport.
Use CurrentSpecReport() instead. You can learn more here: https://onsi.github.io/ginkgo/#getting-a-report-for-the-current-spec
```

fossedihelm replied:
Fixed, It was inherited from upgrading to ginkgo2

iholder-redhat reviewed (commented):
Thanks @fossedihelm!!
This is a great idea! Well done! Happy to see it.

p.s. there are a lot more `tests.BeforeTestCleanup()` calls left, but I guess we can remove the old usage step by step.

iholder-redhat commented:
Thanks a lot @fossedihelm! Great to see this! Awesome work.
/lgtm

fossedihelm commented:
> Thanks @fossedihelm!! This is a great idea! Well done! Happy to see it.
> 
> p.s. there are a lot more `tests.BeforeTestCleanup()` calls left, but I guess we can remove the old usage step by step.

@iholder-redhat Yes, the approach was to starting with a subset of tests and, once we are sure enough, to extend it to all other tests.
I pushed the requested changes, that can be summarized as follow:

- Changed `TestClient` description
- Used `embeded` interfaces, which allows to override the only necessary functions
- Add handler for `Delete` api call
- Changed `Clean()` behavior using `[]CleanableResources`
- Update deprecated function call in `framework.go` file

fossedihelm commented:
/retest

fossedihelm commented:
@iholder-redhat 

- Switched from `Update` to `Patch` in `node.go` https://github.com/kubevirt/kubevirt/blob/cdc7130e8a98cd5384c183cde3636ac3ee74b78b/tests/framework/client/node.go#L76

- Using `Expect()` instead of `panicOnError`

fossedihelm commented:
/retest-required

jean-edouard commented:
/cc

fossedihelm commented:
Rebased

iholder-redhat commented:
> Rebased
re-lgtm
/lgtm

jean-edouard reviewed (commented):
/approve

fossedihelm commented:
/retest

EdDev commented:
This seems like a proposal to change the way we write e2e tests in general.
If this is indeed the case, IMO we need feedback from a wide audience and an improve the exposure of this to the other contributors.

Unlike previous changes to the e2e tests, the PR description hints it is a major logical change.
Could you please present it in some forum and ask for feedback from more maintainers and contributors?

Placing a hold to have a chance to review this and understand the implications. I hope more will join in to review this.
/hold

EdDev reviewed (commented):
I like the general approach of introducing a "framework" to help with our complex e2e tests.
I saw this used in k8s e2e tests and it seemed to have made their tests much more compact and clean.

I would start with a request to introduce this through a design proposal and not directly as a PR.
It affects how tests should be written, and how existing tests could be made cleaner. All this requires clear motivation, goals, structure and phases to be properly introduced.
Other SIG/s need to contribute their own helpers and make sure they can be used, so high level review at the planning stage can be very useful.

To the change proposal itself: I managed to go over only the first commit.
I found it challenging to review, mainly because it introduced many of the client functionality/abstraction without seeing it used. I had no choice but to give up and look at the PR as one change, which made it even harder.
In this regard, I think there needs to be a strategy where the changes are introduced gradually, one functionality at a time, in small portions.

Although I don't feel I even stated to review it from top to bottom, I do see one issue that in k8s community it has been raised several times:
- While a "framework" provides a skeleton and structure to how a test should be written (e.g. abstracting cleanup), the functional helpers (e.g. the kubevirt-client) are a separate part which can be considered as libraries. In practice, what I suggest here is to take out the test-client out of the "framework" and let users use it directly even without the framework.
This will also serve well in introducing it in small commits (and possibly a smaller PR).

Regarding the framework, I am still not deep into the details as I need to fetch locally the PR and review directly. It is too big to be reviewed on GitHub diff (e.g. I tried to look for usages of some methods, but it did not find any, probably because some files are not shown in the diff).
But perhaps I can ask this even before I start: Have you used in this framework GinkGo v2 capabilities (or considered them)?

EdDev reviewed (changes requested):
I started looking at the solution locally and I must admit it is challenging to comprehend.

Would you consider simplifying this with explicit creation methods?
i.e. `f.CreateNamespace(...)`.

With the existing tooling from gingko, the teardown can be as simple as:
```
ns := f.CreateNamespace(...)
DeferCleanup(f.DeleteNamespace, ns)
```

I would strive to simplify everything and make it explicit when used in the tests, not hide it under multiple interfaces.

fossedihelm commented:
@EdDev Thanks for your review and the time spent on this! Few replies inline.

> I like the general approach of introducing a "framework" to help with our complex e2e tests. I saw this used in k8s e2e tests and it seemed to have made their tests much more compact and clean.
> 
> I would start with a request to introduce this through a design proposal and not directly as a PR. It affects how tests should be written, and how existing tests could be made cleaner. All this requires clear motivation, goals, structure and phases to be properly introduced. Other SIG/s need to contribute their own helpers and make sure they can be used, so high level review at the planning stage can be very useful.

Agreed, I will post a design proposal

> To the change proposal itself: I managed to go over only the first commit. I found it challenging to review, mainly because it introduced many of the client functionality/abstraction without seeing it used. I had no choice but to give up and look at the PR as one change, which made it even harder. In this regard, I think there needs to be a strategy where the changes are introduced gradually, one functionality at a time, in small portions.

Can you explain this part? Talking about portions, are you speaking about testClient vs Framework? or smaller portions of testClient, i.e. node/namespace/migrationPolicy/persistentVolume? In the first case I agree with you, in the second one, I think that the test client watching only one resource could results in a not completed functionality creating confusion about which resource is tracked and which is not.
 
> Although I don't feel I even stated to review it from top to bottom, I do see one issue that in k8s community it has been raised several times:
> 
> * While a "framework" provides a skeleton and structure to how a test should be written (e.g. abstracting cleanup), the functional helpers (e.g. the kubevirt-client) are a separate part which can be considered as libraries. In practice, what I suggest here is to take out the test-client out of the "framework" and let users use it directly even without the framework.
>   This will also serve well in introducing it in small commits (and possibly a smaller PR).

I will adjust this PR in the introduction on the only testClient (maybe I will create a new one linking this one), and a design proposal for the framework. WDYT?
 
> Regarding the framework, I am still not deep into the details as I need to fetch locally the PR and review directly. It is too big to be reviewed on GitHub diff (e.g. I tried to look for usages of some methods, but it did not find any, probably because some files are not shown in the diff). But perhaps I can ask this even before I start: Have you used in this framework GinkGo v2 capabilities (or considered them)?

Nope, this PR started before the upgrade to GinkGo v2 and then there was only rebase. I will appreciate any suggestion about it, thanks!

I will address your comment in the new PR, is it ok for you? Thanks

fossedihelm commented on tests/framework/client/client.go:
```diff
@@ -0,0 +1,65 @@
+package client
```
You're right thanks!

fossedihelm commented:
> I started looking at the solution locally and I must admit it is challenging to comprehend.
> 
> Would you consider simplifying this with explicit creation methods? i.e. `f.CreateNamespace(...)`.
> 
> With the existing tooling from gingko, the teardown can be as simple as:
> 
> ```
> ns := f.CreateNamespace(...)
> DeferCleanup(f.DeleteNamespace, ns)
> ```
> 
> I would strive to simplify everything and make it explicit when used in the tests, not hide it under multiple interfaces.

@EdDev I'm not clear what you're referring to: since you are mentioning interfaces I think you are referring to the testClient,
but your example shows usage of the Framework.
The test client was intended on purpose to automatically track changes to the cluster that could generate leftovers.
Here you are suggesting manual handling through `DeferCleanup`.
The two are probably incompatible and so I am beginning to wonder if the testClient in this respect has any use. Or did I misunderstand? Thanks

maiqueb commented on tests/framework/client/node.go:
```diff
@@ -0,0 +1,134 @@
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	v12 "k8s.io/client-go/kubernetes/typed/core/v1"
```
this import path should be changed imo - v12 says nothing informative about the pkg.
